IdentityCore/src/oauth2/MSIDTokenResponseHandler.m

```diff
@@ -0,0 +1,172 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDTokenResponseHandler.h"
+#import "MSIDAADTokenResponse.h"
+#import "MSIDRequestContext.h"
+#import "MSIDAADV1TokenResponse.h"
+#import "MSIDAADV2TokenResponse.h"
+#import "MSIDError.h"
+#import "MSIDClientInfo.h"
+#import "MSIDAADV2RequestParameters.h"
+#import "NSOrderedSet+MSIDExtensions.h"
+
+@implementation MSIDTokenResponseHandler
+
++ (BOOL)processResponse:(MSIDAADTokenResponse *)response
+           refreshToken:(NSString *)refreshToken
+          requestParams:(MSIDRequestParameters *)parameters
+                context:(id<MSIDRequestContext>)context
+                  error:(NSError * __autoreleasing *)error
+{
+    if (!response)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain,
+                                     MSIDErrorInternal, @"processTokenResponse called without a response dictionary", nil, nil, nil, context.correlationId, nil);
+        }
+        return NO;
+    }
+    
+    [self.class checkCorrelationId:response requestCorrelationId:context.correlationId];
+    
+    if (response.error)
+    {
+        if (error)
+        {
+            MSIDErrorCode errorCode = refreshToken ? MSIDErrorServerRefreshTokenRejected : MSIDErrorServerOauth;
+            if ([response isKindOfClass:MSIDAADV2TokenResponse.class])
+            {
+                errorCode = [self.class getErrorCodeForAADV2:response.error];
+            }
+            
+            *error = MSIDCreateError(MSIDOAuthErrorDomain,
+                                     errorCode,
+                                     response.errorDescription,
+                                     response.error,
+                                     nil,
+                                     nil,
+                                     context.correlationId,
+                                     nil);
+        }
+        return NO;
+    }
+    
+    if (!response.accessToken)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Authentication response received without expected accessToken", nil, nil, nil, context.correlationId, nil);
+        }
+        return NO;
+    }
+    
+    if (!response.clientInfo)
```

Comment:
Do we need to fail here?

Reply:
Yes we need to. We return error in this function. Completion block should be executed after error is returned from this function. (Did I understand your question correctly? :) )